client/src/modules/vouchers/complex-voucher.ctrl.js

```diff
@@ -105,6 +111,134 @@ function ComplexJournalVoucherController(Vouchers, $translate, Currencies, Sessi
       });
   }
 
+  // open generic income function
+  function openGenericIncomeTool() {
+    Toolkit.open(genericIncomeTool)
+      .then(function (result) {
+        if (!result) { return; }
+
+        var rows = result.rows;
+        var n = result.rows.length;
+
+        // initial clear
+        vm.Voucher.clear();
+
+        while (n--) {
+          vm.Voucher.addItems(1);
+
+          var lastRowIdx = vm.Voucher.store.data.length - 1;
+          var lastRow = vm.Voucher.store.data[lastRowIdx];
+
+          lastRow.configure(rows[n]);
+        }
+
+        setTimeout(function () {
```

Comment:
This should use the Angular js library [$timeout](https://docs.angularjs.org/api/ng/service/$timeout) which ensures that methods are called within `$apply` etc. 

It would also be useful to leave a comment here explaining why this timeout of 0 seconds is needed!